near/receiver/src/lib.rs

```diff
@@ -0,0 +1,388 @@
+        Price,
+        PriceFeed,
+        PriceIdentifier,
+        Source,
+        Vaa,
+    },
+    std::io::Cursor,
+};
+
+pub mod error;
+pub mod ext;
+pub mod governance;
+pub mod state;
+pub mod tests;
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKeys {
+    Source,
+    Prices,
+    Governance,
+}
+
+/// The `State` contains all persisted state for the contract. This includes runtime configuration.
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Pyth {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    sources: UnorderedSet<Source>,
+
+    /// The Governance Source.
+    gov_source: Source,
+
+    /// The last executed sequence number for governance actions.
+    executed_gov_sequences: UnorderedSet<u64>,
+
+    /// A Mapping from PriceFeed ID to Price Info.
+    prices: UnorderedMap<PriceIdentifier, PriceFeed>,
+
+    /// The `AccountId` of the Wormhole contract used to verify VAA's.
+    wormhole: AccountId,
+
+    /// A hash of the current contract code.
+    ///
+    /// If this `hash` does not match the current contract code, then it indicates that a pending
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    codehash: [u8; 32],
+
+    /// Staleness threshold for rejecting price updates.
+    stale_threshold: Duration,
+
+    /// Fee for updating price.
+    update_fee: u64,
+}
+
+#[near_bindgen]
+impl Pyth {
+    #[init]
+    #[allow(clippy::new_without_default)]
+    pub fn new(
+        wormhole: AccountId,
+        codehash: [u8; 32],
+        initial_source: Source,
+        gov_source: Source,
+        update_fee: u64,
+        stale_threshold: u64,
+    ) -> Self {
+        // Add an initial Source so that the contract can be used.
+        let mut sources = UnorderedSet::new(StorageKeys::Source);
+        sources.insert(&initial_source);
+        Self {
+            prices: UnorderedMap::new(StorageKeys::Prices),
+            executed_gov_sequences: UnorderedSet::new(StorageKeys::Governance),
+            stale_threshold,
+            gov_source,
+            sources,
+            wormhole,
+            codehash,
+            update_fee,
+        }
+    }
+
+    #[init(ignore_state)]
+    pub fn migrate() -> Self {
+        let state: Self = env::state_read().expect("Failed to read state");
+        state
+    }
+
+    /// Instruction for processing VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn update_price_feed(&mut self, vaa_hex: String) -> Result<(), Error> {
```

Comment:
FYI the other contracts now accept a list of VAAs here and for the fee estimate function below. Not sure if it's worth it to change the signature here -- if there's another way to craft an atomic tx that does multiple price updates, then this is fine. Especially so because the callbacks for processing multiple vaas will be gnarly.